spec/asyncapi.md

```diff
@@ -1162,6 +1184,7 @@ Field Name | Type | Description
 <a name="messageObjectHeaders"></a>headers | [Schema Object](#schemaObject) &#124; [Reference Object](#referenceObject) | Schema definition of the application headers. Schema MUST be of type "object". It **MUST NOT** define the protocol headers.
 <a name="messageObjectPayload"></a>payload | `any` | Definition of the message payload. It can be of any type but defaults to [Schema object](#schemaObject). It must match the schema format, including encoding type - e.g Avro should be inlined as either a YAML or JSON object NOT a string to be parsed as YAML or JSON.
 <a name="messageObjectCorrelationId"></a>correlationId | [Correlation ID Object](#correlationIdObject) &#124; [Reference Object](#referenceObject) | Definition of the correlation ID used for message tracing or matching.
+<a name="messageObjectReplyChannel"></a>replyChannel | [ReplyChannel Object](#replyChannelObject) &#124; [Reference Object](#referenceObject) | Should be set if message is used in a [Operation](#operationsObject) using request/reply pattern. If a application creates a inbox for all reponses. The inbox address have to be given in replyChannel field.
```

Comment:
I'm not sure how I feel about having this field here and the `reply` field at the operation level.

What happens if this field says that the reply channel is in the `$message.header#/replyTo` header but the operation says the reply channel is a specific one? Which of the two should we trust? 🤔 

For instance, the following example is ambiguous:

```yaml
operations:
  requestTest:
    channel:
      $ref: '#/channels/requestChannel'
    reply:
      channel:
        $ref: '#/channels/replyChannel'
channels:
  requestChannel:
    address: 'my-request-channel'
    messages:
      req:
        headers: ...
        payload: ...
        replyChannel:
          location: '$message.header#/replyTo'
  replyChannel:
    address: 'my-reply-channel'
```

If we trust `operation.reply` object, the reply channel is always the one with the 'my-reply-channel' address. If we trust the `message.replyChannel` object, the reply channel is dynamic and can't be known in advance.

Am I missing something?

Reply:
It have to be:
```
  replyChannel:
    address: null
```

Otherwise it would be contradictory.
But in general  message.replyChannel is more specific and should be prepared.

Reply:
Yeah, so what I suspected. It leaves the door open to multiple confusions. It's not mentioned anywhere that the channel address would have to be `null`; even if we mention it, it's not intuitive.

So my proposal is that we either go for specifying all this info in the `operation.reply` object or specify everything as part of the message and get rid of `operation.reply`. Would be great to get the pros and cons of each approach.

Reply:
### Today
```
operations:
  requestStatic:
    channel:
      $ref: '#/channels/requestChannelStatic'
    reply:
      channel:
        $ref: '#/channels/replyChannelStatic'
  requestDynamic:
    channel:
      $ref: '#/channels/requestChannelDynamic''
    reply:
      channel:
        $ref: '#/channels/replyChannelDynamic'
channels:
  requestChannelStatic:
    address: 'my-request-channel'
    messages:
      req:
        headers: ...
        payload: ...         
  replyChannelStatic:
    address: 'my-reply-channel'
    
  requestChannelDynamic:
    address: 'my-request-channel'
    messages:
      req:
        headers: ...
        payload: ...
        replyChannel:
          location: '$message.header#/replyTo'
  replyChannelDynamic:
    address: null
```

**Option A:  Having it all at the message**

```
operations:
  requestStatic:
    channel:
      $ref: '#/channels/requestChannelStatic'
  requestDynamic:
    channel:
      $ref: '#/channels/requestChannelDynamic''

channels:
  requestChannelStatic:
    address: 'my-request-channel'
    messages:
      req:
        headers: ...
        payload: ...
        replyChannel:
          $ref: '#/channels/replyChannelStatic'
  replyChannelStatic:
    address: 'my-reply-channel'
    
  requestChannelDynamic:
    address: 'my-request-channel'
    messages:
      req:
        headers: ...
        payload: ...
        replyChannel:
          location: '$message.header#/replyTo'
```

**Down sides:**
- At `operation` level we have no hint that it is a request/reply pattern
- You don't have a reply channel any more. A `channel` is more than the `address`


**Option B:  Having it all at the operation**

```
operations:
  requestStatic:
    channel:
      $ref: '#/channels/requestChannelStatic'
    reply:
      channel:
        $ref: '#/channels/replyChannelStatic'
  requestDynamic:
    channel:
      $ref: '#/channels/requestChannelDynamic''
    reply:
      channel:
        $ref: '#/channels/replyChannelDynamic
      channelAddressLocation: '$message.header#/replyTo'
channels:
  requestChannelStatic:
    address: 'my-request-channel'
    messages:
      req:
        headers: ...
        payload: ...         
  replyChannelStatic:
    address: 'my-reply-channel'
    
  requestChannelDynamic:
    address: 'my-request-channel'
    messages:
      req:
        headers: ...
        payload: ...
  replyChannelDynamic:
    address: null
```

**Down sides:**
- All messages that are possible for given reply channel need to have the reply channel at the same location.

Reply:
What about something like this?

```yaml
channels:
  root:
    address: '/'
    messages:
      ping:
        $ref: '#/components/messages/ping'
      pong:
        $ref: '#/components/messages/pong'
operations:
  pingPong:
    action: send
    channel: 
      $ref: '#/channels/root'
    messages:
      ping:
        $ref: '#/components/messages/ping'
    reply:
      channel: 
        $ref: '#/channels/root'
      messages: # This is a map of request/reply messages.
        ping: # For the "ping" message...
          $ref: '#/messages/pong' # ...you get a "pong" message.
  dynamicPingPong:
    action: send
    channel: 
      $ref: '#/channels/root'
    messages:
      ping:
        $ref: '#/components/messages/ping'
    reply:
      messages:
        ping:
          location: '$message.header#/replyTo'
```

Not perfect but already solving a few of the downsides you mentioned. What do you think?

Reply:
@fmvilas I add a option C in my comment. Because your sample has a problem:
`operations[x].reply.message` but if have to be: `operations[x].message` because your describe the request not the reply message.

Reply:
@GreenRover About your C option:

> contradictory schemas are possible:

I don't think so, `operation[x].messages` should be (if they are specified) subset of messages defined in "connected" channel, so validator/parser should throw error that `$ref: '#/messages/DOS-NOT-EXIST'` not belongs to the `channel.messages`.

> Is operation[x].messages optional if yes means this all messages of related channel.

I also think that if `operation[x].messages` are empty, operation "perform" logic on all messages defined in corresponding channel, but `operation[x].messages` should be (if they are specified) subset of messages defined in "connected" channel.

@fmvilas I like that structure, that for given request message, operation will response with specified message, but it will broke behaviour that we wanna include in v3 https://github.com/asyncapi/spec/issues/829, so maybe something like:

```yaml
operations:
  pingPong:
    action: send
    channel: 
      $ref: '#/channels/root'
    messages:
      ping:
        $ref: '#/components/messages/ping'
    reply:
      channel: 
        $ref: '#/channels/root'
      messages:
        - request: 
            $ref: '#/messages/pong' # For the "ping" message...
          response:
            $ref: '#/messages/pong' # ...you get a "pong" message.
```

I don't like this 😆 but if we wanna go with `$ref` in all places, I don't have any other idea how to handle that.

Reply:
@magicmatatjahu Whit do you have `request` in `reply` ?
OR do you mean `request` and `response` as keyword? If yes, there can be multiple Reponses be possible.

Reply:
@GreenRover 

Yeah, `request` and `response` is a spec keywords for `reply.messages`, `request` defines "request" message, and `response` defines "response" messages for given request message, as we wanna handle all relationship based on refs in v3 (https://github.com/asyncapi/spec/issues/829) we could use keys of map as "references". 

> If yes, there can be multiple Reponses be possible.

Hmmm 🤔 we can treat `reply.messages[x].response` as list:

```yaml
      messages:
        - request: 
            $ref: '#/messages/ping' # For the "ping" message...
          response:
             - $ref: '#/messages/pong' # ...you get a "pong" message.
```

Reply:
@magicmatatjahu sorry but looks weird to me.

Why not this way:
```yaml
operations:
  pingPong:
    action: send
    channel:
      $ref: '#/channels/root'
    messages:
      ping:
        $ref: '#/messages/ping'
        location: '$message.payload#/replyTo'
    reply:
      channel:
        $ref: '#/channels/root'
      messages:
        pong:
          $ref: '#/messages/pong'
```

Reply:
I am not up to date with this proposal, but I know we need to add channel address locations (something like correlationID). If this location is to be outside the message structure then it can be added to `reply.messages[x].request`:

```yaml
      messages:
        - request: 
            channelAddressLocation: '$message.header#/replyTo'
            message:
              $ref: '#/messages/ping' # For the "ping" message...
          response:
             - $ref: '#/messages/pong' # ...you get a "pong" message.
```

but now `reply` has very complex structure so I don't think so that it's good 😅 I don't know tbh what we should think about it.

Reply:
@GreenRover In your example you don't know for what message you should response `pong` message. When operation "performs" logic on single message I think that is simple to understand, but when you have something like that:

```yaml
operations:
  someOperation:
    action: send
    channel:
      $ref: '#/channels/root'
    messages:
      ping:
        $ref: '#/messages/ping'
        location: '$message.payload#/replyTo'
      pong:
        $ref: '#/messages/pong'
        location: '$message.payload#/replyTo'
    reply:
      channel:
        $ref: '#/channels/root'
      messages:
        pong:
          $ref: '#/messages/pong'
        ping:
          $ref: '#/messages/ping'
```

> this operation has circular refs, I know that but it's only a example of multiple response messages

How is a person reading the documentation supposed to know that for a `ping` message the operation should response a `pong` message and vice versa? Fran added example based on keys of `reply.messages` to make relationship between `operations.messages[x]` and `reply.messages[x]` and due to https://github.com/asyncapi/spec/issues/829 issue we shouldn't make relationship based on labels/keys, but `$ref`.

---

Although, as I now think of your way, we can of course define one message in `operation.messages` and all messages with `reply.messages` are answered with a given message. I have no problem with that. However, if we have several messages in `operation.messages` it means that for those messages a given operation can send any message defined in `reply.messages`. WDYT?

Reply:
just to add, althought I'm not sure it is 100% on topic, it is hard to follow with all those different examples. 

it might be possible that when you specify reply, you can say that there are 2 different messages that can come to the reply channel.

Kraken example from my article is not full, because 2.0 is really hard to use with req/res pattern. With `ping` and `pong`, and also other events, there are actually possible 2 messages in reply. For example:

for broken `ping` event you can get:
```json
{
    "errorMessage": "Unsupported field: 'data' for the given msg type: ping",
    "event": "pingStatus",
    "status": "error"
}
```


for broken `subscribe` event you can get:
```json
{
  "errorMessage": "Malformed request",
  "event": "error"
}
```

I my article, I just solved it with JSON schema, but saying that:
```yml
subscriptionStatus:
      type: object
      oneOf:
        - $ref: '#/components/schemas/subscriptionStatusError'
        - $ref: '#/components/schemas/subscriptionStatusSuccess'
```

which is possible, but then it is only possible with JSON schema, and we of course support more formats.